src/order/antisymmetrization.lean

```diff
@@ -0,0 +1,196 @@
+
+## Main declarations
+
+* `antisymm_rel`: The antisymmetrization relation. `antisymm_rel r a b` means that `a` and `b` are
+  related both ways by `r`.
+* `antisymmetrization α r`: The quotient of `α` by `antisymm_rel r`. Even when `r` is just a
+  preorder, `antisymmetrization α` is a partial order.
+-/
+
+open function order_dual
+
+variables {α β : Type*}
+
+section relation
+variables (r : α → α → Prop)
+
+/-- The antisymmetrization relation. -/
+def antisymm_rel (a b : α) : Prop := r a b ∧ r b a
+
+lemma antisymm_rel_swap : antisymm_rel (swap r) = antisymm_rel r :=
+funext $ λ _, funext $ λ _, propext and.comm
+
+@[refl] lemma antisymm_rel_refl [is_refl α r] (a : α) : antisymm_rel r a a := ⟨refl _, refl _⟩
+
+variables {r}
+
+@[symm] lemma antisymm_rel.symm {a b : α} : antisymm_rel r a b → antisymm_rel r b a := and.symm
+
+@[trans] lemma antisymm_rel.trans [is_trans α r] {a b c : α} (hab : antisymm_rel r a b)
+  (hbc : antisymm_rel r b c) :
+  antisymm_rel r a c :=
+⟨trans hab.1 hbc.1, trans hbc.2 hab.2⟩
+
+instance antisymm_rel.decidable_rel [decidable_rel r] : decidable_rel (antisymm_rel r) :=
+λ _ _, and.decidable
+
+@[simp] lemma antisymm_rel_iff_eq [is_refl α r] [is_antisymm α r] {a b : α} :
+  antisymm_rel r a b ↔ a = b := antisymm_iff
+
+alias antisymm_rel_iff_eq ↔ antisymm_rel.eq _
+
+
+end relation
+
+section is_preorder
+variables (α) (r : α → α → Prop) [is_preorder α r]
+
+/-- The antisymmetrization relation as an equivalence relation. -/
+@[simps] def antisymm_rel.setoid : setoid α :=
+⟨antisymm_rel r, antisymm_rel_refl _, λ _ _, antisymm_rel.symm, λ _ _ _, antisymm_rel.trans⟩
+
+/-- The partial order derived from a preorder by making pairwise comparable elements equal. This is
+the quotient by `λ a b, a ≤ b ∧ b ≤ a`. -/
+def antisymmetrization : Type* := quotient $ antisymm_rel.setoid α r
+
+variables {α}
+
+/-- Turn an element into its antisymmetrization. -/
+def to_antisymmetrization : α → antisymmetrization α r := quotient.mk'
+
+/-- Get a representative from the antisymmetrization. -/
+noncomputable def of_antisymmetrization : antisymmetrization α r → α := quotient.out'
+
+instance [inhabited α] : inhabited (antisymmetrization α r) := quotient.inhabited _
+
+@[elab_as_eliminator]
+protected lemma antisymmetrization.ind {p : antisymmetrization α r → Prop} :
+  (∀ a, p $ to_antisymmetrization r a) → ∀ q, p q :=
+quot.ind
+
+@[elab_as_eliminator]
+protected lemma antisymmetrization.induction_on {p : antisymmetrization α r → Prop}
+  (a : antisymmetrization α r) (h : ∀ a, p $ to_antisymmetrization r a) : p a :=
+quotient.induction_on' a h
+
+@[simp] lemma to_antisymmetrization_of_antisymmetrization (a : antisymmetrization α r) :
+  to_antisymmetrization r (of_antisymmetrization r a) = a := quotient.out_eq' _
+
+end is_preorder
+
+section preorder
+variables {α} [preorder α] [preorder β] {a b : α}
+
+lemma antisymm_rel.image {f : α → β} (hf : monotone f) {a b : α} (h : antisymm_rel (≤) a b) :
```

Comment:
I suggest that you swap the explicit arguments: now `h.image` results in a strange error instead of `monotone ?_f → antisymm_rel (≤) (?_f a) (?_f b)`.